pkg/kubelet/nodelease/controller.go

```diff
@@ -0,0 +1,176 @@
+		return
+	}
+	wait.Until(c.sync, c.renewInterval, stopCh)
+}
+
+func (c *controller) sync() {
+	lease, created := c.backoffEnsureLease()
+	// we don't need to update the lease if we just created it
+	if !created {
+		c.retryUpdateLease(lease)
+	}
+}
+
+// backoffEnsureLease attempts to create the lease if it does not exist,
+// and uses exponentially increasing waits to prevent overloading the API server
+// with retries. Returns the lease, and true if this call created the lease,
+// false otherwise.
+func (c *controller) backoffEnsureLease() (*coordv1beta1.Lease, bool) {
+	var (
+		lease   *coordv1beta1.Lease
+		created bool
+		err     error
+	)
+	sleep := 100 * time.Millisecond
+	for {
+		lease, created, err = c.ensureLease()
+		if err == nil {
+			break
+		}
+		// log error
+		glog.Errorf("failed to ensure node lease exists, error: %v", err)
+		// backoff wait
+		time.Sleep(sleep)
+		sleep = minDuration(2*sleep, 7*time.Second)
+	}
+	return lease, created
+}
+
+// ensureLease creates the lease if it does not exist. Returns the lease and
+// a bool (true if this call created the lease), or any error that occurs.
+func (c *controller) ensureLease() (*coordv1beta1.Lease, bool, error) {
+	lease, err := c.client.Get(c.holderIdentity, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		// lease does not exist, create it
+		lease, err := c.client.Create(c.newLease(nil))
+		if err != nil {
+			return nil, false, err
+		}
+		return lease, true, nil
+	} else if err != nil {
+		// unexpected error getting lease
+		return nil, false, err
+	}
+	// lease already existed
+	return lease, false, nil
+}
+
+// retryUpdateLease attempts to update the lease for maxUpdateRetries,
+// call this once you're sure the lease has been created
+func (c *controller) retryUpdateLease(base *coordv1beta1.Lease) {
+	for i := 0; i < maxUpdateRetries; i++ {
+		if _, err := c.client.Update(c.newLease(base)); err != nil {
+			glog.Errorf("failed to update node lease, error: %v", err)
+			if i > 0 && c.onRepeatedHeartbeatFailure != nil {
+				c.onRepeatedHeartbeatFailure()
+			}
+			continue
+		}
+		return
+	}
+	glog.Errorf("failed %d attempts to update node lease, will retry after %v", maxUpdateRetries, c.renewInterval)
+}
+
+// newLease constructs a new lease if base is nil, or returns a copy of base
+// with desired state asserted on the copy.
+func (c *controller) newLease(base *coordv1beta1.Lease) *coordv1beta1.Lease {
+	var lease *coordv1beta1.Lease
+	if base == nil {
+		lease = &coordv1beta1.Lease{}
+	} else {
+		lease = base.DeepCopy()
+	}
+	// Use the bare minimum set of fields; other fields exist for debugging/legacy,
+	// but we don't need to make node heartbeats more complicated by using them.
+	lease.Name = c.holderIdentity
+	lease.Spec.HolderIdentity = pointer.StringPtr(c.holderIdentity)
+	lease.Spec.LeaseDurationSeconds = pointer.Int32Ptr(c.leaseDurationSeconds)
+	lease.Spec.RenewTime = &metav1.MicroTime{c.clock.Now()}
+	return lease
```

Comment:
Is OwnerReference going to be added in follow-up PR, or a different mechanism will be used to ensure the leases are cleaned up?

Reply:
I'm not sure it would work, OwnerReference currently requires the parent and child to be in the same namespace. I was just going to have NLC clean these up for now.